tutorial/tests/testsuite_helpers.py

```diff
@@ -0,0 +1,371 @@
+
+    def __init__(
+        self,
+        test_outputs: List[TestResult],
+        name: str = "",
+        syntax_error: bool = False,
+        success: bool = False,
+        cell_exec_count: int = 0,
+        solution_body: str = "",
+    ):
+        reveal_solution = cell_exec_count > 2 or success
+        output_config = format_success_failure(syntax_error, success, name)
+        output_cell = ipywidgets.Output()
+
+        # For each test, create an alert box with the appropriate message,
+        # print the code output and display code errors in case of failure
+        with output_cell:
+            custom_div_style = '"border: 1px solid; border-color: lightgray; background-color: #FAFAFA; margin: 5px; padding: 10px;"'
+            display(HTML("<h3 style=>Test results</h3>"))
+            display(
+                HTML(
+                    f"""<div class="alert alert-box {output_config.style}"><h4>{output_config.name}</h4>{output_config.result}</div>"""
+                )
+            )
+
+            if not syntax_error:
+                if len(test_outputs) > 0 and test_outputs[0].stdout:
+                    display(
+                        HTML(
+                            f"""
+                                <h4>&#128073; Code output:</h4>
+                                <div style={custom_div_style}>{test_outputs[0].stdout}</div>
+                            """
+                        )
+                    )
+
+                display(
+                    HTML(
+                        f"""
+                            <h4>&#128073; We tested your solution <code>solution_{name}</code> with {'1 input' if len(test_outputs) == 1 else str(len(test_outputs)) + ' different inputs'}.
+                            {"All tests passed!</h4>" if success else "Below you find the details for each test run:</h4>"}
+                        """
+                    )
+                )
+
+                if not success:
+                    for test in test_outputs:
+                        test_name = test.test_name
+                        if match := re.search(r"\[.*?\]", test_name):
+                            test_name = re.sub(r"\[|\]", "", match.group())
+
+                        display(
+                            HTML(
+                                f"""
+                                    <div style={custom_div_style}>
+                                        <h5>{"&#10004" if test.success else "&#10060"} Test {test_name}</h5>
+                                        {format_long_stdout(filters.ansi.ansi2html(test.stderr)) if not test.success else ""}
+                                    </div>
+                                """
+                            )
+                        )
+
+                if not reveal_solution:
+                    display(
+                        HTML(
+                            "<h4>&#128221; A proposed solution will appear after 3 failed attempts.</h4>"
```

Comment:
This is my only real change suggested: print the remaining failed attempts before showing the solution

```suggestion
                            f"<h4>&#128221; A proposed solution will appear after {3 - cell_exec_count} more failed attempt{'s' if cell_exec_count < 2 else ''}.</h4>"
```